File: frontend/src/types/agent.ts
export type TaskStatus = 'TODO' | 'DOING' | 'DONE' | 'BLOCKED' | 'CANCELLED'
export type TaskPriority = 'URGENT' | 'HIGH' | 'MEDIUM' | 'LOW'
export type ConfirmationAction = 'approve' | 'edit' | 'reject' | 'regenerate'

export interface TaskDraft {
  title: string
  description: string
  category: string | null
  deadline: string | null
  estimated_minutes: number | null
  semantic_importance: number
  impact_score: number
  deadline_score: number
  workload_risk_score: number
  dependency_score: number
  priority_reason: string | null
}

export interface TaskDraftEdit {
  title?: string
  description?: string
  category?: string | null
  deadline?: string | null
  estimated_minutes?: number | null
  user_priority?: TaskPriority | null
}

export interface SubtaskDraft {
  step_key: string
  title: string
  description: string
  order: number
  estimated_minutes: number | null
  deadline: string | null
  depends_on: string[]
  completion_weight: number
}

export interface SubtaskPlan {
  parent_task_id: string
  parent_version: number
  summary: string | null
  items: SubtaskDraft[]
  warnings: string[]
}

export interface SubtaskPlanEdit {
  summary?: string | null
  items: SubtaskDraft[]
}

export type ConfirmationEdits = TaskDraftEdit | SubtaskPlanEdit

export interface ConfirmationOptions {
  edits?: ConfirmationEdits
  feedback?: string
}

export interface PendingAction {
  id: string
  user_id: string
  thread_id: string
  action_type: string
  target_id: string | null
  payload: Record<string, unknown>
  confirmation_status: 'pending' | 'approved' | 'rejected' | 'cancelled'
  idempotency_key: string
  created_at: string
  expires_at: string
}

export interface Task {
  id: string
  user_id: string
  parent_id: string | null
  title: string
  description: string
  category: string | null
  status: TaskStatus
  deadline: string | null
  estimated_minutes: number | null
  actual_minutes: number | null
  ai_priority: TaskPriority | null
  user_priority: TaskPriority | null
  effective_priority: TaskPriority | null
  priority_source: 'ai' | 'user' | null
  urgency_score: number | null
  priority_reason: string | null
  progress: number
  is_ai_generated: boolean
  subtask_order: number | null
  depends_on_task_ids: string[]
  completion_weight: number
  creation_source: 'manual' | 'ai_task_parse' | 'ai_decomposition' | null
  created_at: string
  updated_at: string
  completed_at: string | null
  version: number
}

export interface AgentResponse {
  status: string
  thread_id: string
  message: string
  pending_action: PendingAction | null
  task_draft: TaskDraft | null
  task: Task | null
  tasks: Task[]
  candidates: Task[]
  subtask_plan: SubtaskPlan | null
  subtasks: Task[]
  parent_task: Task | null
  deleted_task_id: string | null
  deleted_task_ids: string[]
  deletion_tasks: Task[]
  parent_tasks: Task[]
}

export interface AgentChatRequest {
  user_id: string
  thread_id: string
  message: string
  timezone: string
  request_id: string
}

export interface AgentConfirmRequest {
  user_id: string
  thread_id: string
  action_id: string
  action: ConfirmationAction
  edits?: ConfirmationEdits
  feedback?: string
}

export interface ChatMessage {
  id: string
  role: 'user' | 'agent'
  content: string
  response?: AgentResponse
}

File: agent-service/app/schemas/agent.py
from enum import Enum

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, model_validator

from app.schemas.audit import PendingAction
from app.schemas.draft import TaskDraft
from app.schemas.task import Task, TaskPriority
from app.schemas.subtask import SubtaskPlan, SubtaskPlanEdit


class ConfirmationAction(str, Enum):
    APPROVE = 'approve'
    EDIT = 'edit'
    REJECT = 'reject'
    REGENERATE = 'regenerate'


class TaskDraftEdit(BaseModel):
    model_config = ConfigDict(extra='forbid')

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    category: str | None = None
    deadline: AwareDatetime | None = None
    estimated_minutes: int | None = Field(default=None, gt=0)
    user_priority: TaskPriority | None = None

    @model_validator(mode='after')
    def require_change(self) -> 'TaskDraftEdit':
        if not self.model_fields_set:
            raise ValueError('At least one edited field must be provided')
        return self


class ConfirmationDecision(BaseModel):
    model_config = ConfigDict(extra='forbid')

    action_id: str = Field(min_length=1)
    action: ConfirmationAction
    edits: TaskDraftEdit | SubtaskPlanEdit | None = None
    feedback: str | None = Field(default=None, max_length=1000)

    @model_validator(mode='after')
    def validate_action_payload(self) -> 'ConfirmationDecision':
        if self.action == ConfirmationAction.EDIT and self.edits is None:
            raise ValueError('edits are required for edit action')
        if self.action != ConfirmationAction.EDIT and self.edits is not None:
            raise ValueError('edits are only allowed for edit action')
        return self


class AgentChatRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    user_id: str = Field(min_length=1)
    thread_id: str = Field(min_length=1)
    message: str = Field(min_length=1)
    timezone: str = Field(default='UTC', min_length=1)
    request_id: str = Field(min_length=1)


class AgentConfirmRequest(ConfirmationDecision):
    user_id: str = Field(min_length=1)
    thread_id: str = Field(min_length=1)


class AgentResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    status: str
    thread_id: str
    message: str
    pending_action: PendingAction | None = None
    task_draft: TaskDraft | None = None
    task: Task | None = None
    tasks: list[Task] = Field(default_factory=list)
    candidates: list[Task] = Field(default_factory=list)
    subtask_plan: SubtaskPlan | None = None
    subtasks: list[Task] = Field(default_factory=list)
    parent_task: Task | None = None
    deleted_task_id: str | None = None
    deleted_task_ids: list[str] = Field(default_factory=list)
    deletion_tasks: list[Task] = Field(default_factory=list)
    parent_tasks: list[Task] = Field(default_factory=list)
